apiserver/facades/client/application/application_unit_test.go

```diff
@@ -1603,23 +1964,34 @@ func (s *ApplicationSuite) TestConsumeIdempotent(c *gc.C) {
 		c.Assert(err, jc.ErrorIsNil)
 		c.Assert(results.OneError(), gc.IsNil)
 	}
-	obtained, ok := s.backend.remoteApplications["hosted-mysql"]
-	c.Assert(ok, jc.IsTrue)
-	c.Assert(obtained, jc.DeepEquals, &mockRemoteApplication{
-		name:           "hosted-mysql",
-		sourceModelTag: coretesting.ModelTag,
-		status:         status.Active,
-		offerUUID:      "hosted-mysql-uuid",
-		offerURL:       "othermodel.hosted-mysql",
-		endpoints: []state.Endpoint{
-			{ApplicationName: "hosted-mysql", Relation: charm.Relation{Name: "database", Interface: "mysql", Role: "provider"}}},
-	})
 }
 
 func (s *ApplicationSuite) TestConsumeFromExternalController(c *gc.C) {
+	ctrl := s.setup(c, state.ModelTypeIAAS)
+	defer ctrl.Finish()
+
 	mac, err := apitesting.NewMacaroon("test")
 	c.Assert(err, jc.ErrorIsNil)
 	controllerUUID := utils.MustNewUUID().String()
+
+	s.backend.EXPECT().RemoteApplication("hosted-mysql").Return(nil, errors.NotFoundf(`saas application "hosted-mysql"`))
+	s.backend.EXPECT().SaveController(crossmodel.ControllerInfo{
+		ControllerTag: names.NewControllerTag(controllerUUID),
+		Alias:         "controller-alias",
+		CACert:        coretesting.CACert,
+		Addrs:         []string{"192.168.1.1:1234"},
+	}, coretesting.ModelTag.Id()).Return(nil, nil)
+	s.backend.EXPECT().AddRemoteApplication(state.AddRemoteApplicationParams{
+		Name:                   "hosted-mysql",
+		OfferUUID:              "hosted-mysql-uuid",
+		URL:                    "othermodel.hosted-mysql",
+		ExternalControllerUUID: controllerUUID,
+		SourceModel:            coretesting.ModelTag,
+		Endpoints:              []charm.Relation{{Name: "database", Interface: "mysql", Role: "provider"}},
+		Spaces:                 []*environs.ProviderSpaceInfo{},
+		Macaroon:               mac,
+	}).Return(nil, nil)
+
```

Comment:
Unfortunately the repetition, to some extent, is required since the `state.AddRemoteApplicationParams` structs vary a fair amount between tests. 

I guess I could have a default base, which I copy and change between tests, to emphasize the differences.